lambda-packages/role_handler/index.js

```diff
@@ -156,11 +156,84 @@ const generateRoleName = exports.generateRoleName = function (logicalResourceId)
   return roleName;
 };
 
-const deleteRole = async function (roleName) {
+const updateRole = async function (roleName, props, oldProps) {
+  const iam = new aws.IAM();
+
+  if (waiter) {
+    // Used by the test suite, since waiters aren't mockable yet
+    iam.waitFor = waiter;
+  }
+
+  console.log(`Updating role ${roleName}...`);
+
+  const {
+    Role
+  } = await iam.getRole({
+    RoleName: props.RoleName
+  }).promise();
+
+  let updateRoleProps = {}
+  if (props.Description) {
+    updateRoleProps.Description = props.Description;
+  }
+  if (props.MaxSessionDuration) {
+    updateRoleProps.MaxSessionDuration = props.MaxSessionDuration;
+  }
+  if (updateRoleProps) {
+    updateRoleProps.RoleName = roleName;
+    await iam.updateRole(updateRoleProps).promise();
+  }
+
+  for (const policy of oldProps.Policies || []) {
+    if (!props.Policies || !props.Policies.map(x => x.PolicyName).includes(policy.PolicyName)) {
+      console.log(`Delete role policy ${policy.PolicyName}...`);
+      await iam.deleteRolePolicy({
+        RoleName: roleName,
+        PolicyName: policy.PolicyName
+      }).promise();
+    }
+  }
+  for (const policy of props.Policies || []) {
+    console.log(`Add role policy ${policy.PolicyName}...`);
+    await iam.putRolePolicy({
+      RoleName: roleName,
+      PolicyName: policy.PolicyName,
+      PolicyDocument: policy.PolicyDocument
+    }).promise();
+  }
+
+  for (const arn of oldProps.ManagedPolicyArns || []) {
+    if (!props.ManagedPolicyArns || !props.ManagedPolicyArns.includes(arn)) {
+      console.log(`Detach role policy ${arn}...`);
+      await iam.detachRolePolicy({
+        RoleName: roleName,
+        PolicyArn: arn
+      }).promise();
+    }
+  }
+  for (const arn of props.ManagedPolicyArns || []) {
+    if (!oldProps.ManagedPolicyArns || !oldProps.ManagedPolicyArns.includes(arn)) {
+      console.log(`Attach role policy ${arn}...`);
+      await iam.attachRolePolicy({
```

Comment:
What if the policy has already been detached? Can we catch this condition instead of raising an uncaught exception?

Reply:
Attaching the same policy twice does not throw an exception, so I think this can be kept as is